### ISSUE TITLE
Option to Replay an Already Played Video - Online/Offline - Save Data Usage

### ISSUE THREAD

**Issue by desmonk** [TeamNewPipe/NewPipe#4344]
#### Describe the feature you want

Certain videos might need replaying them once finished watching. So whatever part of the video is already loaded, be it partial or a complete video, users can replay that part again even without internet unless the app is closed just like a browser.

#### Is your feature request related to a problem? Please describe it

This not a problem but close to it as the app does not allow for replaying already played time of the video and therefore results in more data consumption.

#### Additional context

I don't have any to add right n now.

#### How will you/everyone benefit from this feature?

Data savings is one of the biggest advantage here. Users sometimes don't want to download the whole video. Instead they can replay already played one. Second would be ability to have interruption free "second time viewing" experience. Say for some reason, the connection is really bad and video loads poorly, a user here can be assured that they will be able to watch it once certain amount of video is ready and loaded.

To summarize, this falls short to a complete download but interestingly does the same while keeping what is already played in memory of app until it is closed down.

PS. I search with different keywords to find similar issues but many are broad and there are countless pages. Forgive me again if this is a re-post.

**Comment by opusforlife2** [comment]
When you finish watching a video and see the replay button come up, tapping that replay button will load the video from the cache already.

**Comment by desmonk** [comment]
Reply works but if we manually move the slider back and forth then it restarts downloading kind of forgetting the cache.

**Comment by opusforlife2** [comment]
Try it with internet off.

Play a video until the end. Then turn internet off and tap replay.

**Comment by desmonk** [comment]
Steps to reproduce:
While watching a video, interrupt the playback with a pause. And move slider to rewind and after that the video starts to load again. Yes, that is true if we watch until the end then replay is fine. But that eventually became a offline download in cache itself.

Difference here is when a user pauses a video somewhere and goes back and forth and say does this few times then at one point the loading bar starts again kind of forgetting the cache in the process.

**Comment by opusforlife2** [comment]
Hmmm. That seems like a cache size limitation. This means we need to download videos in a temp folder as they stream. Then I'm considering this a duplicate of #478. Please add your suggestion as a comment there.

**Comment by desmonk** [comment]
Alright. I will add there. Thank you!

**Comment by TheDcoder** [comment]
@opusforlife2 
> Then I'm considering this a duplicate of #478.

In my opinion this issue is not a strict duplicate of #478, this is a feature request related to offline replay and the other is a feature request for the ability to play videos in exoplayer (NewPipe's player), both of them are related but not the same.

Once #478 is implemented, this feature can be implemented by adding an option to temporarily download all streamed videos in a cache folder or something similar.

**Comment by TheDcoder** [comment]
Looks like this issue might be a duplicate of #3932, #478 was also recently updated to mention that issue in the top post.

**Comment by opusforlife2** [comment]
Actually, I think I'll open #3932 (as it is broader in scope than this one) and insert #478 as a prerequisite.
